drivers: modem: cleanup

mike-scott opened the pull request:
- WNC-M14A2A requires use of GPIOs in the driver.  Let's make sure we imply GPIO if the modem is selected, as several HW now do not select GPIO by default.
- Cleanup the drivers/modem/Kconfig due to left over artifact from the WNC-M14A2A split.
- Fix timeout handling for WNC-M14A2A
- WNC-M14A2A: Cleaned up includes

mike-scott committed:
drivers: modem: wnc-m14a2a: imply GPIO usage

CONFIG_GPIO may not be enabled on some HW, so let's make sure to
enable it if the dependencies are met, as the WNC-M14A2A driver
needs to set power, reset and other GPIOs.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: remove leftover artifact from wncm14a2a split

In commit c1f24abb1381 ("drivers: modem: wistron kconfig separation"),
A single line of the WNC-M14A2A congigs was left in the main modem
Kconfig.

Let's clean up the Kconfig file by removing the left over.

Signed-off-by: Michael Scott <mike@foundries.io>

ulfalizer commented on drivers/modem/Kconfig.wncm14a2a:
```diff
@@ -10,6 +10,7 @@ menuconfig MODEM_WNCM14A2A
 	bool "Enable Wistron LTE-M modem driver"
 	select MODEM_RECEIVER
 	select NET_OFFLOAD
+	select GPIO
```
(Maybe this could be a `depends on` as well, but `GPIO` has no dependencies at least.)

mike-scott replied:
@ulfalizer Ugh, I forget: does using select this way *not* include the dependencies for GPIO (if it had any?).  I would prefer to have the driver select dependencies if it needs them, rather than placing that logic on the samples or boards that select the driver.

ulfalizer replied:
`select` forces symbols on, ignoring dependencies. It's risky for that reason.

I wrote up a section on `select` pitfalls and recommendations on the [Kconfig - Tips and Best Practices](https://docs.zephyrproject.org/latest/guides/kconfig/index.html#select-statements) page.

mike-scott replied:
@ulfalizer Right.  I remember that.  So, `select` has a very narrow use-case where it's only safe to use it when it's in a small set of options that are all known good configurations.  In place of that, we're now adding `depends on` throughout the  Kconfig options and deferring the logic needed to enable drivers into the HW and/or the sample configurations.   And, there was a push to remove the defaults from the HW configurations recently, which leaves only the samples.

... This doesn't seem like it's going end well.

Doesn't the Linux kernel have a mechanism to accomplish what we are doing?

mike-scott replied:
Could we introduce a PR test which checks to see if configs which are `select`ed have dependencies and if so: rejects the change?  There should be some basic configs which drivers can select (without fear of future breakage): ADC, AUDIO, GPIO, I2C, PWM, etc ...

ulfalizer replied:
I wonder if such a check might be too broad.

For example, say you have a bunch of symbols for audio codecs that are all defined next to each other, and that all have the same (say complex) dependencies. In that case, it might not be horrible to have a symbol like `CONFIG_ALL_FOO_CODECS` next to them that selects all of the FOO-related ones.

There are some legitimate and reasonably safe uses for `select` like that. It's mostly when people start treating it as "just another `depends on` alternative" that things tend to get messy real quick. :)

Check out the [imply](https://docs.zephyrproject.org/latest/guides/kconfig/index.html#the-imply-statement) statement too. It enables symbols only if their dependencies are satisfied.

mike-scott replied:
I'll switch to `imply` usage here and repost.

mike-scott commented:
Changes:
- Replaced `select` with `imply` and tested compile for `nrf52840_pca10056` which does not select GPIO by default.
- Updated commit description with removal of the word "select" (instead saying "enable if dependencies are met").

mike-scott committed:
drivers: modem: wncm14a2a: Fix timeout handling for AT@SOCKCONN

Socket-based API sends timeout K_FOREVER(-1) to net_context_connect()
function where previously net_app APIs used an actual timeout value.
Now that we've switched to socket-based APIs, we reveal poor handling
for the timeout value which causes an error in the WNC-M14A2A driver
due to math performed on the -1 value.

Let's be sure to check for valid timeout values prior to performing
math and then make sure it falls within the allowable range of values
for the AT@SOCKCONN command (30 seconds to 360 seconds).

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Added a bugfix for WNC-M14A2A modem connection timeout which was revealed during NET_OFFLOAD testing (this has been broken for some time).

mike-scott committed:
drivers: modem: wncm14a2a: remove left-over includes

Commit fd1401495b93 ("net/tcp: Move net_tcp_set_hdr() away from net
core") removed <net/tcp.h>.  This results in a not found error being
generated when CONFIG_NET_TCP is enabled.

We also, don't need to include <net/udp.h> anymore due to new
net_pkt API changes.

Let's clean up the includes by removing both of them.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Added another bugfix for WNC-M14A2A cleaning up the include files (TCP include file was removed causing a build break).

tbursztyka reviewed (approved):
lgtm
